src/Model/Behavior/FootprintBehavior.php

```diff
@@ -68,22 +68,56 @@ public function initialize(array $config)
     }
 
     /**
-     * Injects configured fields into finder conditions.
-     *
-     * @param \Cake\Event\Event $event Event.
-     * @param \Cake\ORM\Query $query Query.
-     * @param \ArrayObject $options Options.
+     * {@inheritDoc}
+     */
+    public function implementedEvents()
+    {
+        /* map all configured events to a single handler */
+        return array_map(
+            function () {
+                return 'dispatch';
+            },
+            $this->_config['events']
+        );
+    }
+
+    /**
+     * Dispatch an event to the corresponding function
+     * Called by the event system as instructed by implementedEvents()
+     * @param Event $event Event.
+     * @param Query|EntityInterface $data Query or Entity.
+     * @param ArrayObject $options Options.
      * @return void
      */
-    public function beforeFind(Event $event, Query $query, ArrayObject $options)
+    public function dispatch(Event $event, $data, ArrayObject $options)
     {
         $eventName = $event->name();
         if (empty($this->_config['events'][$eventName])) {
             return;
         }
 
-        $config = $this->_config['events'][$eventName];
-        foreach (array_keys($config) as $field) {
+        $fields = $this->_config['events'][$eventName];
+
+        if ($data instanceof EntityInterface) {
```

Comment:
It is your project and you will have to maintain it, so I will do it your way. That said, I disagree 100%: Neither this code nor the methods called are affected by the event type, they just do their specific work on a Query or Entity. You will have more code to read, and to maintain, e.g. when event names change, and no added flexibility or robustness. The code already throws an exception when incompatible events are used.